My Site Dashboard (Phase 2): Hook up Quick Action buttons

momo-ozawa opened the pull request:
Part of #17874 

## Description
- Hooked up buttons on the Quick Actions cell on the blog dashboard
   - Stats -> Show stats screen
   - Posts -> Show post list screen
   - Media -> Show media library screen
   - Pages -> Show page screen
- Added event tracking for the Quick Action buttons 

## How to test

Make sure to run your sandbox with the 1.1 API diff applied and to have the My Site Dashboard flag enabled:

1. Go to My Site
2. Go to the dashboard, tap on the **Stats** button
3. ✅ The app should navigate to the Stats screen
4. ✅ The following event should be tracked:`🔵 Tracked: stats_accessed <blog_id: 1, site_type: blog, tap_source: dashboard>`
5. Go back to the dashboard, tap on the **Posts** button
6. ✅ The app should navigate to the Post List screen
7. ✅ The following event should be tracked:`🔵 Tracked: site_menu_opened <blog_id: 1, menu_item: posts, site_type: blog, tap_source: dashboard>`
8. Go back to the dashboard, tap on the **Media** button
9. ✅ The app should navigate to the Media Library screen
10. ✅ The following event should be tracked:`🔵 Tracked: site_menu_opened <blog_id: 1, menu_item: library, site_type: blog, tap_source: dashboard>`
11. Go back to the dashboard, tap on the **Pages** button
12. ✅ The app should navigate to the Page List screen
13. ✅ The following event should be tracked:`🔵 Tracked: site_menu_opened <blog_id: 1, menu_item: pages, site_type: blog, tap_source: dashboard>`

https://user-images.githubusercontent.com/6711616/153590538-deb18cc9-2f13-4e75-8243-31033852c407.mp4


## Regression Notes
1. Potential unintended areas of impact
N/A

2. What I did to test those areas of impact (or what existing automated tests I relied on)
N/A

3. What automated tests I added (or what prevented me from doing so)
N/A

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Commit:
Add: hook up quick action buttons

Commit:
Add: event tracking for quick action buttons

Commit:
Refactor: extract Quick Actions extension to it's own file

leandroalonso commented on WordPress/Classes/ViewRelated/Blog/Blog Dashboard/BlogDashboardViewController+QuickActions.swift:
```diff
@@ -0,0 +1,52 @@
+
+extension BlogDashboardViewController {
+
+    func showStats() {
```
Do you think we can move that to the `DashboardQuickActionsCardCell`?

The cell has the `Blog` and the `ViewController` injected in it, I feel we can have this logic there.

I like this approach because:

* It decouples the cell from `BlogDashboardViewController`
* It makes this cell reusable anywhere else, as long as a blog and a viewController is provided
* We avoid similar issues we have a few days ago with `BlogDetailsViewController` (in which a lot of logic was tied to this VC)

momo-ozawa replied:
Fixed! f55e5b3

leandroalonso commented on WordPress/Classes/ViewRelated/Blog/Blog Dashboard/BlogDashboardViewController+QuickActions.swift:
```diff
@@ -0,0 +1,52 @@
+        controller.navigationItem.largeTitleDisplayMode = .never
+        showDetailViewController(controller, sender: self)
+
+        QuickStartTourGuide.shared.visited(.stats)
```
I wonder if we can create a static method on `StatsViewController` that: given a `Blog` and a `ViewController`, it presents itself and also handle the `QuickStartTourGuide.shared.visited(.stats)`.

Otherwise, we'll have to make use of the `QuickStartTourGuide.shared.visited(.stats)` in every place we present it.

Ps.: this is to "finish" the tour for checking stats, right?

momo-ozawa replied:
Fixed! edc0ac1

> Ps.: this is to "finish" the tour for checking stats, right?

Yep  -- may have been a bit too early to add this before tackling the quick start task though

leandroalonso commented on WordPress/Classes/ViewRelated/Blog/Blog Dashboard/BlogDashboardViewController+QuickActions.swift:
```diff
@@ -0,0 +1,52 @@
+    func showPostList() {
+        trackQuickActionsEvent(.openedPosts)
+
+        let controller = PostListViewController.controllerWithBlog(blog)
```
Same as before.

We can have a `PostListViewController.present(for: blog, in: controller)`

momo-ozawa replied:
Fixed! https://github.com/wordpress-mobile/WordPress-iOS/commit/edc0ac107363c8675579937984d107989a8260db

leandroalonso reviewed (commented):
@momo-ozawa left a few comments, feature-wise, all good!

I was worried that in the Simulator I wasn't able to scroll through the buttons (the button captures the touch event and I wasn't able to scroll). However, I tested on a real device and it works just fine. :)

Commit:
Refactor: move button actions into cell

pachlava commented:
Took me some time to set the environment up, but it works as you described! Events tracking and navigation works with no issues. The only thing I found so far is:

1. When creating a new site, the previously selected will still be active once creation is over. Not sure if this is related to the PR.
2. If `Dashboard` was active, and a new site was created (and the user replied with `Show me around`), the dashboard will look like a mixture of `Site Menu` and `Dashboard` (and still the old site being active):

<img width="341" alt="Screenshot 2022-02-11 at 19 15 46" src="https://user-images.githubusercontent.com/73365754/153638687-841ccc42-b7f9-4731-922b-aa5ead283249.png">

(it fixes itself as soon the user switches to `Site Menu` and back).

Will test a bit more today.

pachlava commented:
Again, both of the points I mention above are most probably not related directly to the PR... I've missed several most recent PRs.

Commit:
Refactor: define static methods on quick action destination vcs

Definte static methods on StatsVC, PostListVC, MediaLibraryVC, PagesListVC to push themselves on to a source view controller and also handle Quick Start visits.

leandroalonso commented:
@pachlava ~good catch! Is not related to this PR indeed, I added to this issue: https://github.com/wordpress-mobile/WordPress-iOS/issues/17873~

(Momo had already added)

leandroalonso reviewed (approved):
:shipit:
